lean_source/07_Topology/source_01_Filters.lean

```diff
@@ -0,0 +1,465 @@
+example : 𝓝 (x₀, y₀) = 𝓝 x₀ ×ᶠ 𝓝 y₀ := nhds_prod_eq
+-- QUOTE.
+
+/- TEXT:
+This operation is defined in term of the pull-back operation and the ``inf`` operation:
+``F ×ᶠ G = (comap prod.fst F) ⊓ (comap prod.snd G)``.
+Here the ``inf`` operation refers to the lattice structure on ``filter X`` for any type ``X``:
+``F ⊓ G`` is the greatest filter that is smaller than bother ``F`` and ``G``, generalizing the notion
+of the intersection of sets.
+
+A lot of proofs in  mathlib use all this structure (``map``, ``comap``, ``inf``, ``sup``, ``prod``) 
+to give algebraic proofs about convergence without ever referring to members of filters.
+You can practice this on the following lemma, unfolding the definition of ``tendsto``
+and ``filter.prod`` if needed.
+TEXT. -/
+
+-- QUOTE:
+#check le_inf_iff
+
+example (f : ℕ → ℝ × ℝ) (x₀ y₀ : ℝ) : 
+  tendsto f at_top (𝓝 (x₀, y₀)) ↔ 
+  tendsto (prod.fst ∘ f) at_top (𝓝 x₀) ∧ tendsto (prod.snd ∘ f) at_top (𝓝 y₀) :=
+sorry
+-- QUOTE.
+
+-- SOLUTIONS:
+
+example (f : ℕ → ℝ × ℝ) (x₀ y₀ : ℝ) : 
+  tendsto f at_top (𝓝 (x₀, y₀)) ↔ 
+  tendsto (prod.fst ∘ f) at_top (𝓝 x₀) ∧ tendsto (prod.snd ∘ f) at_top (𝓝 y₀) :=
+calc
+tendsto f at_top (𝓝 (x₀, y₀)) ↔ map f at_top ≤ 𝓝 (x₀, y₀) : iff.rfl
+... ↔ map f at_top ≤ 𝓝 x₀ ×ᶠ 𝓝 y₀ : by rw nhds_prod_eq
+... ↔ map f at_top ≤ (comap prod.fst (𝓝 x₀) ⊓ comap prod.snd (𝓝 y₀)) : iff.rfl
+... ↔ map f at_top ≤ comap prod.fst (𝓝 x₀) ∧ map f at_top ≤ (comap prod.snd (𝓝 y₀)) : le_inf_iff
+... ↔ map prod.fst (map f at_top) ≤ 𝓝 x₀ ∧ map prod.snd (map f at_top) ≤ 𝓝 y₀ : 
+        by rw [← map_le_iff_le_comap, ← map_le_iff_le_comap]
+... ↔ map (prod.fst ∘ f) at_top ≤ 𝓝 x₀ ∧ map (prod.snd ∘ f) at_top ≤ 𝓝 y₀ : by rw [map_map, map_map]
+/- begin
+  rw nhds_prod_eq,
+  unfold tendsto filter.prod,
+  rw [le_inf_iff, ← map_le_iff_le_comap, map_map, ← map_le_iff_le_comap, map_map]
+end -/
+
+/- TEXT:
+The ordered type ``filter X`` is actually a *complete* lattice: there is a bottom element and a top element and
+any set of filters on ``X`` have an ``Inf`` and a ``Sup``.
+
+Note that in the definition of filters, given the second property
+(if ``U`` belong to ``F`` then anything larger also belongs), the first property 
+(the set of all inhabitants of ``X`` belongs to ``F``) is
+equivalent to asking that ``F`` is not the empty collection of sets.
+This shouldn't be confused with a more subtle question. The 
+definition does not prohibit ``∅ ∈ F``. If ``∅ ∈ F`` then
+``∀ U : set X, U ∈ F``, hence ``F`` is a
+rather trivial filter. This trivial filter is precisely the bottom element of the complete
+lattice ``filter X``.
+This contrasts with the definition of filters in
+Bourbaki who doesn't allow filters containing the empty set. 
+Because of this choice, we need an
+extra assumption of non-triviality in some lemmas, but gain
+nicer global formal properties for the theory. We've seen already that it gives us a
+bottom element. It also allowed us to define ``principal : set X → filter X`` without
```

Comment:
```suggestion
bottom element. It also allows us to define ``principal : set X → filter X`` without
```

Reply:
I don't understand why you suggest this changed. The definition of `principal` indeed appeared before in that file.

Reply:
It's a very minor point, I think stylistically present tense tends to sound better than past tense even if it was indeed earlier in the file, but it's yeah, very minor, leave it if you like it, it's not grammatically incorrect.